[WIP]Thunderfield

Filatelele opened the pull request:
Немного WIP, немного говнокод, но это не суть. 

Суть - что-то типо свалки, где игрок может пострелять и устроить дуэли на баллонах. Любителям пострелять больше не нужно дожидаться рольки, чтобы ~~устроить говно~~. 
Работает так - игрок платит 50 кредитов и получает 5 респавнов на тандере. Убийство другого игрока приносит очки, которые можно потратить в отвратительной копипасте аплинка на оружие, броню и всякий другой мусор. 
Автомат можно взломать емагом и получить 10 очков на каждый респавн, нужно быть осторожным - читы вредят здоровью и приводят к брейнроту.


Что сделано на данный момент: 
- [x] VR тела для игроков
- [x] Набросан код игровых подов

TODO list:
- [x] Почистить копипасту и говнокод
- [x] Замаппить игровой зал на станции
- [x] FEATURE - в синий код из подов выбрасывает всех, кроме ассистентов. 
- [x] Сделать карту тандерфилда на Z2 уровне
- [x] Debug?




- [ ] FEATURE - процедурно генерируемая карта тандерфилда(А нужно ли оно?)

<details> 
  <summary>Скриншоты </summary>

![](http://joxi.ru/YmE3qaqs0LknRA.jpg)

</details>

:cl:
 - rscadd[link]:  Battle-royale тандерфилд

Filatelele committed:
Thunderfield

Filatelele commented on code/game/objects/items/devices/uplinks.dm:
```diff
@@ -204,3 +202,72 @@ A list of items and costs is stored under the datum of every game mode, alongsid
 	hidden_uplink = new(src)
 	hidden_uplink.uses = 20
+
+/obj/item/device/uplink/hidden/vr_uplink //Uplink for thunderfield
```
Отвратительная копипаста, простите меня, мейнтейнеры, я не смог придумать ничего лучше

LudwigVonChesterfield replied:
Время идёт, а копипаста. Копипаста не меняется.

Filatelele commented on code/game/turfs/unsimulated/floor.dm:
```diff
@@ -13,3 +13,25 @@
+/turf/unsimulated/floor/self_cleaning
+	var/list/uncleanable_items = list(/mob/living, /obj/item/thunder_dog_tag, /mob/dead/observer, /obj/effect/landmark, /obj/item/weapon) //We dont want this items to be deleted
+
+/turf/unsimulated/floor/self_cleaning/proc/cleaner() //Not sure if this is the correct way to do that
```
Как понятно из комментариев - я не совсем уверен, насколько это легитно делать так

AgRevol commented:
А ты запилил рандомные карты для него?

Filatelele commented:
Мне это даже в голову не пришло, хех, надо будет запилить

Filatelele committed:
Merge branch 'master' of https://github.com/GammaStation/Gamma-Station into Thundergame

# Conflicts:
#	maps/z2.dmm

Filatelele committed:
Карта?

Filatelele committed:
Fixes

Process() >> power_change(), убрал ненужный !

DieSiege commented:
Пущай при синем коде важных для станции ребят выплевывает.

Kurshan commented on code/_onclick/hud/other_mobs.dm:
```diff
@@ -136,3 +136,54 @@
 	popup.set_content(dat)
 	popup.open()
 
+/datum/hud/proc/vrhuman_hud(ui_style = 'icons/mob/screen_vrhuman.dmi')
```
Вынеси в отдельный файл вот это вот всё

Kurshan commented on code/game/objects/items/devices/uplinks.dm:
```diff
@@ -204,3 +202,72 @@ A list of items and costs is stored under the datum of every game mode, alongsid
+/obj/item/device/uplink/hidden/vr_uplink/Topic(href, href_list)
+	..()
+
+	world.log << "active?"
```
Ну ты понял.

LudwigVonChesterfield replied:
15 дней спустя подтверждаю актуальность.

Kurshan commented on code/game/turfs/unsimulated/floor.dm:
```diff
@@ -13,3 +13,25 @@
+		if(is_type_in_list(A, uncleanable_items))
+			continue
+		qdel(A)
+	addtimer(CALLBACK(src, .proc/cleaner), CLEANING_DELAY)
```
Уже Зве намекал, что лучше промышлять таким не на постоянной основе. Я со своей стороны настоятельно рекомендую повесить это на кнопку, которую будут жмякать между раундами егроки. На бее так и сделано, емнип.

Kurshan commented on code/game/turfs/unsimulated/floor.dm:
```diff
@@ -13,3 +13,25 @@
+/turf/unsimulated/floor/self_cleaning/proc/cleaner() //Not sure if this is the correct way to do that
+	for(var/atom/A in contents)
+		if(is_type_in_list(A, uncleanable_items))
+			continue
```
мелочь, но зачем континью, когда можно логическое отрицание сделать?

Kurshan commented on code/modules/thunderfield2562/thunderfield.dm:
```diff
@@ -0,0 +1,32 @@
+	if(!isvrhuman(user))
+		return
+	user.mind.thunder_points += points
+	to_chat(user, "<span class='warning'>You get [points] points! Your total points are: [user.mind.thunder_points]</span>")
```
При каждом подъеме монеток спамить об этом? Может лучше кнопочку или ещё какой уведомлятор?

Kurshan commented on code/modules/thunderfield2562/thunderfield.dm:
```diff
@@ -0,0 +1,32 @@
+
+/obj/item/thunder_dog_tag/atom_init()
+	. = ..()
+	if(!points)
```
Я пока не вижу, где они могут быть проинициализированы

Kurshan commented on code/modules/thunderfield2562/thunderfield_gamepod.dm:
```diff
@@ -0,0 +1,222 @@
+
+//Code for attackby
+
+/obj/machinery/gamepod/attackby(obj/item/I, mob/user)
```
Копипаста на копипасте. Такое ощущение что даже не смотрел в то, что копируешь-вставляешь

Kurshan commented on code/modules/thunderfield2562/thunderfield_gamepod.dm:
```diff
@@ -0,0 +1,222 @@
+	D.money -= transaction_amount
+	station_account.money += transaction_amount
+	var/datum/transaction/T = new()
+	T.target_name = "[station_account.owner_name] (via [src.name])"
```
Нет, серьезно, сделай прок там и вынеси вот эти все действия. Дали вам возможность создания самопальных проков - не хочу, хочу говно жрать

Kurshan commented on code/modules/thunderfield2562/thunderfield_people.dm:
```diff
@@ -0,0 +1,126 @@
+	var/died = FALSE                                    //Look death() proc here, for comments
+	var/obj/item/device/uplink/hidden/vr_uplink/vr_shop //To buy stuff
+	alpha = 127
+	see_in_dark = 8 //Just because there is no light on arena
```
А оно разве не апдейтится в лайф проке?

Kurshan commented on code/modules/thunderfield2562/thunderfield_people.dm:
```diff
@@ -0,0 +1,126 @@
+
+/mob/living/carbon/human/vrhuman/proc/generate_random_body()
+
+	var/obj/randomcatcher/CATCH = new /obj/randomcatcher(src)
```
Классные названия переменных

Kurshan commented on code/modules/thunderfield2562/thunderfield_people.dm:
```diff
@@ -0,0 +1,126 @@
+		return
+	died = TRUE
+	var/obj/item/thunder_dog_tag/dog_tag = new /obj/item/thunder_dog_tag/(loc) //Spawning dog tag for players to get points
+	dog_tag.points = 1
```
Сперва в инициализации самого объекта выставляем один, потом тут, а потом ещё раз при наличии майнда.

Kurshan reviewed (changes requested):
Честно, эту копипасту даже читать уже не хочется, ты бы сперва хоть сам прошёлся по тому, что наделал. Я половину моментов уже подмечать не стал, оставлю это на потом.

Filatelele committed:
Some fixes

LudwigVonChesterfield commented on code/_onclick/hud/_defines.dm:
```diff
@@ -93,6 +93,7 @@
 #define ui_health "EAST-1:28,CENTER:17"
 #define ui_internal "EAST-1:28,CENTER+1:19"
 
+
```
Невероятно важное изменение касающееся темы ПРа.

LudwigVonChesterfield commented on code/datums/mind.dm:
```diff
@@ -69,6 +69,12 @@
 	var/spent_TC = 0
 
+	//This is here for thunderfield (aka VR)
+	var/thunder_points = 0 //Points for buying stuff on thunderfield
```
Даже поменяв тело, небо, аллаха, прожав респавн, кол-во моих очков сохранится. Я не прав?

Filatelele replied:
>Прожав респавн
После прожатия респавна у тебя новый майнд создается, вроде как. Или я не прав? 
>Поменяв тело 
Ну, если у тебя переносят мозг в другое тело - то очки сохраняются, в чем проблема то

LudwigVonChesterfield commented on code/datums/uplinks_items.dm:
```diff
@@ -785,3 +785,189 @@
+/proc/get_thunderfield_items()
+	for(var/item in typesof(/datum/thunderfield_item))
+		var/datum/thunderfield_item/I = new item()
+		if(!I.item) //That can happen?!
```
Да, потому-что typesof(/datum/thunderfield_item) включает сам класс /datum/thunderfield_item, у которого это переменная - нулл. Я бы смотрел в сторону subtypesof() ;).

Filatelele committed:
Merge remote-tracking branch 'remotes/upstream/master' into Thundergame

# Conflicts:
#	icons/obj/items.dmi
#	maps/z2.dmm

Filatelele committed:
Get rid of trash

Filatelele committed:
Изменен способ очистки пола

Переделал процесс очистки пола - теперь он вызывается самими егроками при необходимости с КД в 600 тиков, чего, думаю, вполне достаточно.
Так же подчистил некоторое количество лишнего мусора

Filatelele committed:
Fixes

Filatelele commented:
Короче, это пока-что еще WIP, так как необходимо перелопатить немного копипасты и добавить пару фич.

Filatelele committed:
От синего кода и выше играть могут только гражданские профки

Filatelele committed:
Fixes

Filatelele committed:
Игровой зал на Z1

Filatelele committed:
Map fix

Filatelele committed:
Z2

Filatelele committed:
Update floor.dm

Filatelele committed:
Fixes!

Filatelele committed:
Нам нужно больше фиксов

Filatelele committed:
FIXES!11

DieSiege commented:
Понеслась душа в рай.
